pkg/operation/botanist/controlplane/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,572 @@
+	Labels map[string]string
+	// HealthSyncPeriod describes the duration of how often the health of existing resources should be synced
+	HealthSyncPeriod *time.Duration
+	// KubeConfig configures the gardener-resource-manager to target another cluster for creating resources.
+	// If this is not set resources are created in the cluster the gardener-resource-manager is deployed in
+	KubeConfig *component.Secret
+	// LeaseDuration configures the lease duration for leader election
+	LeaseDuration *time.Duration
+	// MaxConcurrentHealthWorkers configures the number of worker threads for concurrent health reconciliation of resources
+	MaxConcurrentHealthWorkers *int32
+	// RenewDeadline configures the renew deadline for leader election
+	RenewDeadline *time.Duration
+	// ResourceClass is used to filter resource resources
+	ResourceClass *string
+	// RetryPeriod configures the retry period for leader election
+	RetryPeriod *time.Duration
+	// SyncPeriod configures the duration of how often existing resources should be synced
+	SyncPeriod *time.Duration
+	// TargetDisableCache disables the cache for target cluster and always talk directly to the API server (defaults to false)
+	TargetDisableCache *bool
+	// WatchedNamespace restricts the gardener-resource-manager to only watch ManagedResources in the defined namespace.
+	// If not set the gardener-resource-manager controller watches for ManagedResources in all namespaces
+	WatchedNamespace *string
+}
+
+func (r *ResourceManager) Deploy(ctx context.Context) error {
+	if err := r.ensureServiceAccount(ctx); err != nil {
+		return err
+	}
+
+	if err := r.ensureRBAC(ctx); err != nil {
+		return err
+	}
+
+	if err := r.ensureService(ctx); err != nil {
+		return err
+	}
+
+	if err := r.ensureDeployment(ctx); err != nil {
+		return err
+	}
+
+	return r.ensureVPA(ctx)
+}
+
+func (r *ResourceManager) Destroy(ctx context.Context) error {
+	objectsToDelete := []client.Object{
+		r.emptyVPA(),
+		r.emptyDeployment(),
+		r.emptyService(),
+		r.emptyServiceAccount(),
+	}
+	role, err := r.emptyRoleInWatchedNamespace()
+	if err == nil {
+		objectsToDelete = append(objectsToDelete, role)
+	}
+	rb, err := r.emptyRoleBindingInWatchedNamespace()
+	if err == nil {
+		objectsToDelete = append(objectsToDelete, rb)
+	}
+	cr, err := r.emptyClusterRole()
+	if err == nil {
+		objectsToDelete = append(objectsToDelete, cr)
+	}
+	crb, err := r.emptyClusterRoleBinding()
+	if err == nil {
+		objectsToDelete = append(objectsToDelete, crb)
+	}
+
+	return kutil.DeleteObjects(
+		ctx,
+		r.client,
+		objectsToDelete...,
+	)
+}
+
+func (r *ResourceManager) ensureRBAC(ctx context.Context) error {
+	deployResourcesIntoAnotherCluster := r.values.KubeConfig != nil
+	if deployResourcesIntoAnotherCluster {
+		if r.values.WatchedNamespace == nil {
+			if err := r.ensureClusterRole(ctx, allowManagedResources); err != nil {
+				return err
+			}
+			if err := r.ensureClusterRoleBinding(ctx); err != nil {
+				return err
+			}
```

Comment:
I also do not see a use case for it currently.
My motivation: The resource_manager package exposes the configuration `watchedNamespace` and `kubeconfig`. I thought how should it react if the kubeconfig is set but not the watchedNamespace.
An alternative solution would be to throw an error when calling deploy, like `If kubeconfig is set there always needs to be be a watchedNamespace configured`. Would you prefer that over the above implementation?

Reply:
I see, thanks for explaining. I guess when you wanted to follow this approach consequently, you'd also need to support the 4th case i.e., `kubeConfig == nil &&  WatchedNamespace != nil`, which is not covered yet, right?

While I get your point, I also see the benefits to support solely the cases we need right now because when you read the code, it's not quite obvious which case is used where and less branches usually mean better readability. WDYT?

Reply:
The case you point out above is covered. If `kubeconfig == nil` then it does not matter if `watchedNamespace` is set or not. The grm always needs a ClusterRole with full access to deploy into the whole cluster.

I agree with your observation that less code is always better and prematurely including code for other use cases which are not yet on the horizon is bad. I did a commit on another branch to see how it looks if I remove the above and add validation(test adaption missing). See https://github.com/gardener/gardener/commit/8ab692eef1f9807e3a69efd2c08d67e82b30b642

After looking at it I would personally leave it like it is right now as I do not see a major gain in readability with validation over the current implementation. For me the validation adds other questions, like why is this from the standpoint of the grm package a non supported usecase.
WDYT?

Reply:
I just realized I would need to validate in the deploy function because the kubeconfig can be set after the instance got created. This would need to be adapted in the above commit. Anyways, my point from above still stands.

Reply:
Okay, thanks for checking and explaining.